### DISCUSSION TITLE
How can I configure Ocelot programmatically?

### DISCUSSION THREAD

**Discussion by cizu64** [ThreeMammals/Ocelot#2217] [Q&A] [answered]
## I get this error message when trying to configure ocelot programmatically

```
Ocelot.Responder.Middleware.ResponderMiddleware[0]
      requestId: 0HN8B1C7OOUES:00000001, previousRequestId: No PreviousRequestId,
message: 'Error Code: UnableToFindDownstreamRouteError 
Message: Failed to match Route configuration for upstream path: /banks, verb: GET.
errors found in ResponderMiddleware. Setting error response for request path:/banks, request method: GET'
```

## This is my code
```csharp
using System.Security.Claims;
using Microsoft.AspNetCore.Mvc.Filters;
using Ocelot.Configuration.File;
using Ocelot.Configuration.Repository;

public class Middleware : IMiddleware
{
    private readonly IFileConfigurationRepository _fileConfigurationRepository;
    public Middleware(IFileConfigurationRepository fileConfigurationRepository)
    {
        _fileConfigurationRepository = fileConfigurationRepository;
    }
    private List<FileRoute> Configuration(string downStreamPathTemplate)
    {
       
    }
    public async Task InvokeAsync(HttpContext context, RequestDelegate next)
    {
       
        var fileConfig = new FileConfiguration
        {
            Routes = new List<FileRoute>
            {
                new FileRoute
                {
                    DownstreamPathTemplate = "/api/bank/banks",
                    DownstreamScheme = "https",
                    DownstreamHostAndPorts = new List<FileHostAndPort>
                    {
                        new FileHostAndPort { Host = "https://gateway.mybankapp123.com", Port = 443 }
                    },
                    DownstreamHttpMethod = "GET",
                    UpstreamPathTemplate = "/banks",
                    UpstreamHttpMethod = new List<string> { "GET" }
                }
            },
            GlobalConfiguration = new FileGlobalConfiguration
            {
                BaseUrl = "https://localhost:7174"
            }
        };


        await _fileConfigurationRepository.Set(fileConfig);
        
        await next.Invoke(context);
    }
}
```

## This is my program.cs
```csharp
builder.Services.AddOcelot();
builder.Services.AddScoped<Middleware>();
var app = builder.Build();

app.UseHttpsRedirection();
app.UseAuthentication();
app.UseAuthorization();
//await app.UseOcelot().ContinueWith(a => app.UseMiddleware<Middleware>());
await app.UseOcelot();
app.UseMiddleware<Middleware>();
app.Run();
```

**Comment by raman-m** [comment]
## 😄 🐯 
Yes, you can! And everybody can! Yahoo!
**My final answer** → https://github.com/ThreeMammals/Ocelot/discussions/2217#discussioncomment-11372021

**Reply by cizu64** [reply]
Should i mark this as answer?

**Reply by cizu64** [reply]
I get it! I changed the host to google because i don't want to give the real host.

**Reply by raman-m** [reply]
**Final answer** &rarr; https://github.com/ThreeMammals/Ocelot/discussions/2217#discussioncomment-11372021

**Comment by raman-m** [comment]
OK Let's start professional discussion...

> ```csharp
>     private readonly IFileConfigurationRepository _fileConfigurationRepository;
>     public Middleware(IFileConfigurationRepository fileConfigurationRepository)
>     {
>         _fileConfigurationRepository = fileConfigurationRepository;
>     }
> ```
> 

The proposed middleware solution is based on the `IFileConfigurationRepository` interface, which is implemented by the `FileConfigurationRepository` class &rarr;
https://github.com/ThreeMammals/Ocelot/blob/a79ae8a29ed899e855d2ddf322dfae5738b96f0f/src/Ocelot/DependencyInjection/OcelotBuilder.cs#L79
This service within the DI-container is solely responsible for file operations. This is the correct direction; however, the configuration logic in Ocelot is rather complex. First, you must conduct some research on the Ocelot code by examining [all references to IFileConfigurationRepository](https://github.com/search?q=repo%3AThreeMammals%2FOcelot%20IFileConfigurationRepository&type=code)! The appropriate solution should be grounded in the functionality of `FileConfigurationController`:
https://github.com/ThreeMammals/Ocelot/blob/a79ae8a29ed899e855d2ddf322dfae5738b96f0f/src/Ocelot/Configuration/FileConfigurationController.cs#L38-L55
To change the configuration dynamically after the Ocelot application starts, it is necessary to utilize both the `IFileConfigurationRepository` and `IFileConfigurationSetter` services.

**Reply by raman-m** [reply, marked as answer]
### The second,
Ocelot has no fluent syntax to build configuration on the fly.
But injecting `FileConfiguration` object is possible during app start in the [ConfigurationBuilderExtensions](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/DependencyInjection/ConfigurationBuilderExtensions.cs) class methods:
https://github.com/ThreeMammals/Ocelot/blob/a79ae8a29ed899e855d2ddf322dfae5738b96f0f/src/Ocelot/DependencyInjection/ConfigurationBuilderExtensions.cs#L165-L170
This method will deserialize the object to disk.
If you want to keep configuration in memory, then this method has `MergeOcelotJson` option:
https://github.com/ThreeMammals/Ocelot/blob/a79ae8a29ed899e855d2ddf322dfae5738b96f0f/src/Ocelot/DependencyInjection/ConfigurationBuilderExtensions.cs#L185-L190

### So, final solution based on .NET 8
#### Program.cs (implicit usings)
```csharp
using Ocelot.Configuration.File;
using Ocelot.DependencyInjection;
using Ocelot.Middleware;

var builder = WebApplication.CreateBuilder(args);

// Build Ocelot's configuration object on the fly: add routes, global config, etc.
var ocelotConfig = new FileConfiguration(); // ... initialize props: add routes, global config, etc.

builder.Configuration
    .SetBasePath(builder.Environment.ContentRootPath)
    .AddJsonFile($"appsettings.json", true, true)
    .AddEnvironmentVariables()

    // !!! Solution
    .AddOcelot(ocelotConfig) // MergeOcelotJson.ToFile : writing config JSON back to disk
    .AddOcelot(ocelotConfig, builder.Environment, MergeOcelotJson.ToMemory); // merging to memory

builder.Services.AddOcelot();

var app = builder.Build();
await app.UseOcelot();
await app.RunAsync();
```
Hope it helps!

**Reply by cizu64** [reply]
Thanks a lot!

**Reply by raman-m** [reply]
Just additional links to our official docs where injection of `FileConfiguration` is explained:
- [Dependency Injection](https://ocelot.readthedocs.io/en/latest/features/dependencyinjection.html) &rarr; [`IConfigurationBuilder` extensions](https://ocelot.readthedocs.io/en/latest/features/dependencyinjection.html#iconfigurationbuilder-extensions) &rarr; [`AddOcelot` methods](https://ocelot.readthedocs.io/en/latest/features/dependencyinjection.html#addocelot-methods) &rarr; [Signatures with `FileConfiguration` parameter explained](https://ocelot.readthedocs.io/en/latest/features/dependencyinjection.html#addocelot-methods:~:text=Signatures%20of%20the%20versions%20to%20indicate%20the), and read **Note 2**

However, developers typically do not read documentation regularly 😄 and it appears that we need to add brief subsection after this section:
- [Configuration](https://ocelot.readthedocs.io/en/latest/features/configuration.html) &rarr; [Merging Configuration Files](https://ocelot.readthedocs.io/en/latest/features/configuration.html#merging-configuration-files) &rarr; [Merging files to memory](https://ocelot.readthedocs.io/en/latest/features/configuration.html#merging-files-to-memory)

...something named as "_Write dynamic configurations on the fly_" or "_Injecting `FileConfiguration` objects_"

**Reply by raman-m** [reply]
Answer in updated docs for Configuration feature &rarr; https://github.com/ThreeMammals/Ocelot/commit/cf8daf7becacb5f91f134ed912b4520e72322a48
